plivo.py

```diff
@@ -20,12 +21,37 @@ class PlivoError(Exception):
     pass
 
 
-def validate_signature(uri, post_params, signature, auth_token):
-    for k, v in sorted(post_params.items()):
-        uri += k + v
-    return base64.encodestring(hmac.new(auth_token, uri, sha1).digest()).strip() == signature
-
-
+def validate_request_signature(uri, signature, auth_token, params=None):
+    """
+    Validates requests made by Plivo to your servers.
+    See https://www.plivo.com/docs/xml/request/#validation
+
+    :param uri: Your server URL
+    :param params: POST Parameters passed to your URL, in case of POST request. Will be ignored if URL contains a query
+    string
+    :param signature: X-Plivo-Signature header
+    :param auth_token: Plivo Auth token
+    :return: True if the request matches signature, False otherwise
+    """
+    parsed_uri = urlparse(uri.encode('utf-8'))
+    qs = parsed_uri.query
+    if qs:
+        # get params from query string
+        all_params = dict(parse_qsl(qs, keep_blank_values=True))
+        # remove parameters from query string
+        encoded_request = urljoin(uri, parsed_uri.path).encode('utf-8')
+    else:
+        all_params = params.copy() if params else dict()
+        encoded_request = uri.encode('utf-8')
+    for k, v in sorted(all_params.items()):
+        encoded_key = k.encode('utf-8')
+        if isinstance(v, basestring):
+            encoded_val = v
+        elif v:
+            encoded_val = str(v)
```

Comment:
This is wrong if the assumption is that `v` can be something else than a string. If `v` can only be a string, then the `elif` is redundant.

The reason this code is wrong for non-strings is that `v` can be falsey (e.g. `0`, `False`, `None`, `[]`, etc.) and such case `encoded_val` is undefined (or takes the value of the previous iteration of this loop):

```
In [32]: all_params = { 'CallUUID': 'fake-call-uuid', 'Duration': 0, 'CallStatus': 'ringing' }

In [33]: all_params.items()
Out[33]: [('Duration', 0), ('CallUUID', 'fake-call-uuid'), ('CallStatus', 'ringing')]

In [34]: encoded_request = 'https://myserver.com/plivo_call/'

In [35]: for k,v in all_params.items():
    ...:     encoded_key = k.encode('utf-8')
    ...:     if isinstance(v, basestring):
    ...:         encoded_val = v
    ...:     elif v:
    ...:         encoded_val = str(v)
    ...:     encoded_request += encoded_key + encoded_val
    ...:
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-35-d5ef27fce03f> in <module>()
      5     elif v:
      6         encoded_val = str(v)
----> 7     encoded_request += encoded_key + encoded_val
      8

NameError: name 'encoded_val' is not defined
```